Tests/GRPCNIOTransportHTTP2Tests/HTTP2TransportTLSEnabledTests.swift

```diff
@@ -0,0 +1,402 @@
+        else {
+          Issue.record(
+            "Should be a NIOSSLExtraError.failedToValidateHostname error, but was: \(String(describing: rootError.cause))"
+          )
+          return false
+        }
+
+        return true
+      }
+    )
+  }
+
+  @Test("Error is surfaced when server fails client verification")
+  // Verification should fail because the server does not have trust roots containing the client cert.
+  func testServerFailsClientValidation() async throws {
+    await #expect(
+      performing: {
+        try await self.executeUnaryRPCForEachTransportPair { security in
+          [
+            HTTP2TransportTLSEnabledTests.Transport(
+              server: .posix(
+                .tls(
+                  .mTLS(
+                    certificateChain: [.bytes(security.server.certificate, format: .der)],
+                    privateKey: .bytes(security.server.key, format: .der)
+                  )
+                )
+              ),
+              client: .posix(
+                .tls(
+                  .mTLS(
+                    certificateChain: [.bytes(security.client.certificate, format: .der)],
+                    privateKey: .bytes(security.client.key, format: .der)
+                  ) {
+                    $0.trustRoots = .certificates([
+                      .bytes(security.server.certificate, format: .der)
+                    ])
+                    $0.serverHostname = "localhost"
+                  }
+                )
+              )
+            )
+          ]
+        }
+      },
+      throws: { error in
+        guard let rootError = error as? RPCError else {
+          Issue.record("Should be an RPC error")
+          return false
+        }
+        #expect(rootError.code == .unavailable)
+        #expect(
+          rootError.message
+            == "The server accepted the TCP connection but closed the connection before completing the HTTP/2 connection preface."
+        )
+
+        guard
+          let sslError = rootError.cause as? NIOSSL.BoringSSLError,
+          case .sslError = sslError
+        else {
+          Issue.record(
+            "Should be a NIOSSL.sslError error, but was: \(String(describing: rootError.cause))"
+          )
+          return false
+        }
+
+        return true
+      }
+    )
+  }
+}
+
+struct TestSecurity {
+  struct Server {
+    let certificate: [UInt8]
+    let key: [UInt8]
+  }
+
+  struct Client {
+    let certificate: [UInt8]
+    let key: [UInt8]
+  }
+
+  let server: Server
+  let client: Client
+
+  init() throws {
+    let server = try Self.createSelfSignedDERCertificateAndPrivateKey(name: "Server Certificate")
+    let client = try Self.createSelfSignedDERCertificateAndPrivateKey(name: "Client Certificate")
+
+    self.server = Server(certificate: server.cert, key: server.key)
+    self.client = Client(certificate: client.cert, key: client.key)
+  }
+
+  private static func createSelfSignedDERCertificateAndPrivateKey(
```

Comment:
naming nit: factory functions should be "make" per https://www.swift.org/documentation/api-design-guidelines/